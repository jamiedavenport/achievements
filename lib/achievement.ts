import * as yup from "yup";

export type AchievementDto = {
  when: Date;
  what: string;
};

export const achievementSchema = yup.object().shape({
  when: yup.date().required(),
  what: yup.string().required(),
});

export type Achievement = {
  id: string;
  owner: string;
  when: Date;
  what: string;
};

export type CreateAchievement = (
  achievement: Omit<Achievement, "id">
) => Promise<Achievement>;

export type ListAchievements = (owner: string) => Promise<Achievement[]>;

export type AchievementRepo = {
  createAchievement: CreateAchievement;
  listAchievements: ListAchievements;
};
