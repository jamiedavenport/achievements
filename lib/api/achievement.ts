import { Achievement, AchievementDto } from "lib/achievement";

const parseAchievement = (json: Record<string, string>): Achievement => {
  return {
    owner: json.owner,
    what: json.what,
    when: new Date(json.when),
    id: json.id,
  };
};

export const createAchievement = async (values: AchievementDto) => {
  const resp = await fetch(`/api/achievements`, {
    method: "POST",
    body: JSON.stringify(values),
  });

  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error);
  }

  return parseAchievement(await resp.json());
};

export const fetchAchievements = async () => {
  const resp = await fetch(`/api/achievements`);

  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error);
  }

  return (await resp.json()).map(parseAchievement);
};
