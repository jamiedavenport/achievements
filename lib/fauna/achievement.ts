import { AchievementRepo, CreateAchievement } from "lib/achievement";
import fauna, { query } from "faunadb";
import { faunaSecret } from "lib/config";
import formatISO from "date-fns/formatISO";

const client = new fauna.Client({ secret: faunaSecret });
const q = query;

const createAchievement: CreateAchievement = async (achievement) => {
  const when = formatISO(achievement.when, { representation: "date" });

  // TODO: Improve Fauna typing
  const resp = await client.query<any>(
    q.Create(q.Collection("achievements"), {
      data: {
        ...achievement,
        when,
      },
    })
  );

  return {
    id: resp.ref.id,
    ...achievement,
  };
};

export const achievementRepo: AchievementRepo = {
  createAchievement,
};
