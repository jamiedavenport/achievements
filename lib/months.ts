import { Achievement } from "./achievement";
import getMonth from "date-fns/getMonth";

export const months = [
  {
    id: 0,
    label: "January",
  },
  {
    id: 1,
    label: "February",
  },
  {
    id: 2,
    label: "March",
  },
  {
    id: 3,
    label: "April",
  },
  {
    id: 4,
    label: "May",
  },
  {
    id: 5,
    label: "June",
  },
  {
    id: 6,
    label: "July",
  },
  {
    id: 7,
    label: "August",
  },
  {
    id: 8,
    label: "September",
  },
  {
    id: 9,
    label: "October",
  },
  {
    id: 10,
    label: "November",
  },
  {
    id: 11,
    label: "December",
  },
].reverse();

export const filterAchievements = (
  achievements: Achievement[]
): Record<number, Achievement[]> => {
  const buckets: Record<number, Achievement[]> = {};
  for (let i = 0; i < 12; i++) {
    buckets[i] = [];
  }

  achievements.forEach((achievement) => {
    const month = getMonth(achievement.when);
    buckets[month].push(achievement);
  });

  return buckets;
};
