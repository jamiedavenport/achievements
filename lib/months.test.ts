import { Achievement } from "./achievement";
import { filterAchievements } from "./months";

describe("filterAchievements", () => {
  it("should put events into month buckets", () => {
    const owner = "me";
    const achievements: Achievement[] = [
      {
        when: new Date(2021, 0, 1),
        what: "Something in Jan",
        owner,
        id: "1",
      },
      {
        when: new Date(2021, 1, 1),
        what: "Something in Feb",
        owner,
        id: "2",
      },
      {
        when: new Date(2021, 11, 1),
        what: "Something in Dec",
        owner,
        id: "3",
      },
      {
        when: new Date(2021, 0, 31),
        what: "Something else in Jan",
        owner,
        id: "4",
      },
      {
        when: new Date(2021, 2, 31),
        what: "Birthday!",
        owner,
        id: "5",
      },
      {
        when: new Date(2021, 3, 1),
        what: "April fools",
        owner,
        id: "6",
      },
    ];

    const buckets = filterAchievements(achievements);
    expect(buckets).toMatchSnapshot();
  });
});
