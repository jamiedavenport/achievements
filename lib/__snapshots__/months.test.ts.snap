// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterAchievements should put events into month buckets 1`] = `
Object {
  "0": Array [
    Object {
      "id": "1",
      "owner": "me",
      "what": "Something in Jan",
      "when": 2021-01-01T00:00:00.000Z,
    },
    Object {
      "id": "4",
      "owner": "me",
      "what": "Something else in Jan",
      "when": 2021-01-31T00:00:00.000Z,
    },
  ],
  "1": Array [
    Object {
      "id": "2",
      "owner": "me",
      "what": "Something in Feb",
      "when": 2021-02-01T00:00:00.000Z,
    },
  ],
  "10": Array [],
  "11": Array [
    Object {
      "id": "3",
      "owner": "me",
      "what": "Something in Dec",
      "when": 2021-12-01T00:00:00.000Z,
    },
  ],
  "2": Array [
    Object {
      "id": "5",
      "owner": "me",
      "what": "Birthday!",
      "when": 2021-03-30T23:00:00.000Z,
    },
  ],
  "3": Array [
    Object {
      "id": "6",
      "owner": "me",
      "what": "April fools",
      "when": 2021-03-31T23:00:00.000Z,
    },
  ],
  "4": Array [],
  "5": Array [],
  "6": Array [],
  "7": Array [],
  "8": Array [],
  "9": Array [],
}
`;
